TUTORIAL.adoc

```diff
@@ -0,0 +1,537 @@
+The `strategy` namespace also contains the mechanisms for combining rewrite steps.
+For now it is the only namespace we need concern ourselves with.
+Pretty much anything we could do with the other namespaces, we can achieve with a `rewrite`.
+
+[source,clojure]
+----
+(ns meander.examples
+  (:require [meander.delta.strategy :as m]))
+----
+
+
+== Our first rewrite
+
+[source,clojure]
+----
+((m/rewrite
+  ?x
+  "foo")
+ "bar")
+;=> "foo"
+----
+
+Rewrite takes a match pattern and and a substitution pattern.
+In this case the match pattern is `?x` and the substitution pattern is `"foo"`.
+`?x` is a logic variable. In this pattern it will match with anything, as it stands alone.
+
+Meander's `rewrite` is a macro that constructs a function.
+Meander is compiling an efficient rewrite function for us to use with arbitrary inputs.
+Here we immediately invoke our rewriting function on the input `"bar"`.
+The result of calling the rewrite function on `"bar"` is `"foo"`.
+
+Using the rewrite immediately like this is a convenient way to try our rewrite expressions as we go.
+
+
+== Reusing rewrite
+
+Once we are happy with a transformation, it makes sense to give it a name.
+
+[source,clojure]
+----
+(def foo-bar
+  (m/rewrite
+    ?x
+    "foo"))
+
+(foo-bar "bar")
+;=> "foo"
+(foo-bar "baz")
+;=> "baz"
+----
+
+Meander compiles a function that takes some input, and rewrites it to a new output.
+
+
+== Structural pattern matching
+
+Unlike Clojure's pervasive sequence abstraction that applies to all collections,
+Meander matches structure explicitly.
+
+[source,clojure]
+----
+(def g
+  (m/rewrite
+    [?a ?b ?c ?d]
+    {?a ?c ?b ?d}))
+
+(g [1 2 3 4])
+;=> {1 3, 2 4}
+(g '(1 2 3 4))
+;=> *fail*
+----
+
+If we look for a vector, we will only match with a vector.
+When there is a match, the output is explicitly of the type that the pattern specifies.
+
+
+== Example
+
+[source,clojure]
+----
+(def not-not
+  (m/rewrite
+    (not (not ?x))
+    ?x))
+
+(not-not
+  (not (not 1)))
+;=> 1
+----
+
+
+== Combining steps
+
+Here's where things get really interesting.
+Meander has a `bottom-up` combinator which allows us to rewrite terms anywhere in an expression.
+The idea is that the expression is treated as a tree, and Meander will explore that tree for us until it finds a pattern match.
+There is also an `until` combinator that keeps applying the same transform.
+We'll put these combinators into a handy macro, as we'll be using it in a few places.
+
+[source,clojure]
+----
+(defmacro bup [& body]
+  `(m/until = (m/bottom-up (m/rewrite ~@body))))
+----
+
+Combinators allow us to control how rewriting should occur.
+In this case we have specified that `bup` expressions should be applied bottom up until the expression stabilizes.
+Our `bup` expects patterns just like `rewrite`, and will produce a function, just like `rewrite`.
+
+[source,clojure]
+----
+(def nn
+  (bup
+    (not (not ?x))
+    ?x))
+
+(nn (and (not (not (or 1 2)) (or 3 (not (not 4))))
```

Comment:
There are a couple of missing close parens:
```
(nn (and (not (not (or 1 2))) (or 3 (not (not 4)))))
```
Also same problem as above - the argument evaluates to `3` before it's even passed into `nn`

Reply:
Will fix.